unison-sync/src/Unison/Sync/Types.hs

```diff
@@ -0,0 +1,515 @@
+  { entities :: Map Hash (Entity HashJWT Hash Text)
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON DownloadEntitiesResponse where
+  toJSON (DownloadEntitiesResponse entities) =
+    object
+      [ "entities" .= entities
+      ]
+
+data UpdatePathRequest = UpdatePathRequest
+  { path :: RepoPath,
+    expectedHash :: Maybe TypedHash, -- Nothing requires empty history at destination
+    newHash :: TypedHash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON UpdatePathRequest where
+  toJSON (UpdatePathRequest path expectedHash newHash) =
+    object
+      [ "path" .= path,
+        "expected_hash" .= expectedHash,
+        "new_hash" .= newHash
+      ]
+
+instance FromJSON UpdatePathRequest where
+  parseJSON = Aeson.withObject "UpdatePathRequest" $ \obj -> do
+    path <- obj .: "path"
+    expectedHash <- obj .: "expected_hash"
+    newHash <- obj .: "new_hash"
+    pure UpdatePathRequest {..}
+
+-- | Not used in the servant API, but is a useful return type for clients to use.
+data UpdatePathResponse
+  = OutOfDate OutOfDateHash
+  | MissingDependencies (NeedDependencies Hash)
+  deriving stock (Show, Eq, Ord)
+
+data NeedDependencies hash = NeedDependencies
+  { missingDependencies :: NESet hash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON hash => ToJSON (NeedDependencies hash) where
+  toJSON (NeedDependencies missingDependencies) =
+    object ["missing_dependencies" .= missingDependencies]
+
+instance (FromJSON hash, Ord hash) => FromJSON (NeedDependencies hash) where
+  parseJSON = Aeson.withObject "NeedDependencies" $ \obj -> do
+    missingDependencies <- obj .: "missing_dependencies"
+    pure NeedDependencies {..}
+
+data OutOfDateHash = OutOfDateHash
+  { repoPath :: RepoPath,
+    expectedHash :: Maybe TypedHash,
+    actualHash :: Maybe TypedHash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON OutOfDateHash where
+  toJSON (OutOfDateHash repoPath expectedHash actualHash) =
+    object
+      [ "repo_path" .= repoPath,
+        "expected_hash" .= expectedHash,
+        "actual_hash" .= actualHash
+      ]
+
+instance FromJSON OutOfDateHash where
+  parseJSON = Aeson.withObject "OutOfDateHash" $ \obj -> do
+    repoPath <- obj .: "repo_path"
+    expectedHash <- obj .: "expected_hash"
+    actualHash <- obj .: "actual_hash"
+    pure OutOfDateHash {..}
+
+data UploadEntitiesRequest = UploadEntitiesRequest
+  { repoName :: RepoName,
+    entities :: Map Hash (Entity TypedHash TypedHash Text)
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON UploadEntitiesRequest where
+  toJSON (UploadEntitiesRequest repoName entities) =
+    object
+      [ "repo_name" .= repoName,
+        "entities" .= entities
+      ]
+
+instance FromJSON UploadEntitiesRequest where
+  parseJSON = Aeson.withObject "UploadEntitiesRequest" $ \obj -> do
+    repoName <- obj .: "repo_name"
+    entities <- obj .: "entities"
+    pure UploadEntitiesRequest {..}
+
+data Entity hash optionalHash text
```

Comment:
Not necessary for this PR, but it would be nice to document this a bit, esp
 why we landed on two different type variables for hash things, `hash` and `optionalHash`.

Reply:
I switched the name to `replacementHash` which is hopefully clear from context since it's only used in the Patch constructor 😄